Add functional tests for KeystoneAPI

kajinamit opened the pull request:
This introduces the functional tests framework and initial test coverage about KeystoneAPI webhook and controller.

kajinamit commented:
I've submitted the PR to update the prow job https://github.com/openshift/release/pull/40912

kajinamit commented on tests/functional/base_test.go:
```diff
@@ -0,0 +1,82 @@
+	return instance.Status.Conditions
+}
+
+func GetCronJob(name types.NamespacedName) *batchv1.CronJob {
```
This can be later moved to lib-common

rabi commented on controllers/keystoneapi_controller.go:
```diff
@@ -699,6 +699,13 @@ func (r *KeystoneAPIReconciler) reconcileNormal(ctx context.Context, instance *k
+			condition.RequestedReason,
+			condition.SeverityInfo,
+			condition.DeploymentReadyRunningMessage))
+		return ctrl.Result{RequeueAfter: time.Second * 10}, fmt.Errorf("No pod is available")
```
I think you can just do  `return ctrl.Result{}, nil` here as this controller is watching the deployment resource, once the deploy resource changes, this will reconcile again..

gibizer commented on tests/functional/suite_test.go:
```diff
@@ -0,0 +1,177 @@
+	kclient, err := kubernetes.NewForConfig(cfg)
+	Expect(err).ToNot(HaveOccurred(), "failed to create kclient")
+
+	err = (&keystonev1.KeystoneAPI{}).SetupWebhookWithManager(k8sManager)
+	Expect(err).NotTo(HaveOccurred())
```
I'm wondering if we need  to wait for the webhook server to become ready before we start the test case. E.g. https://github.com/openstack-k8s-operators/nova-operator/blob/97ce65fd52b42c6dd2350b69082da389ae62088e/test/functional/suite_test.go#L235-L245

kajinamit replied:
I haven't seen the issue caused by webhook server non-ready but it makes sense to add that wait. I've updated the change to include it.

gibizer reviewed (approved):
Thanks you for taking the time adding the envtest coverage for the change. I have couple of nits inline, nothing blocking as they can be fixed in a follow up.

kajinamit commented on tests/functional/keystone_controller_test.go:
```diff
@@ -0,0 +1,464 @@
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	When("A Keystone instance is created", func() {
```
Fixed.

kajinamit commented on tests/functional/keystone_controller_test.go:
```diff
@@ -0,0 +1,464 @@
+			DeferCleanup(th.DeleteInstance, CreateKeystoneAPI(keystoneApiName, GetDefaultKeystoneAPISpec()))
+		})
+
+		It("should have the Spec fields initialized", func() {
```
Updated accordingly

kajinamit commented:
> Thanks you for taking the time adding the envtest coverage for the change. I have couple of nits inline, nothing blocking as they can be fixed in a follow up.

Thank you, @gibizer, for taking time for reviewing this change. ! Addressed these points.

kajinamit commented:
/retest

gibizer commented on Makefile:
```diff
@@ -116,8 +116,8 @@ golangci-lint:
-test: manifests generate fmt vet envtest ## Run tests.
-	KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) --bin-dir $(LOCALBIN) -p path)" go test ./... -coverprofile cover.out
+test: manifests generate fmt vet envtest ginkgo ## Run tests.
+	KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) --bin-dir $(LOCALBIN) -p path)" $(GINKGO) --trace --cover --coverpkg=../../pkg/keystone,../../controllers,../../api/v1beta1 --coverprofile cover.out --covermode=atomic ${PROC_CMD} $(GINKGO_ARGS) ./tests/...
```
If you add PROC_CMD the the ginkgo command line then I guess you also want to set it to something. In nova-operator we calculate it with:
```
PROCS?=$(shell expr $(shell nproc --ignore 2) / 2)
PROC_CMD = --procs ${PROCS}
```

kajinamit replied:
Ahh, I didn't notice these and was wondering why we need that PROC_CMD. Fixed the command.

gibizer reviewed (approved):
Looks good. Thank you Takash!

stuggi commented:
@kajinamit could you please rebase this one?

stuggi reviewed (approved):
/lgtm
